torch/distributed/_tensor/sharding_prop.py

```diff
@@ -288,16 +290,26 @@ def spec_to_strategy(spec: object) -> object:
                     ):
                         expected_input_spec_list = []
                         for idx, arg_spec in enumerate(arg):
-                            if arg_spec.placements != out_spec_list[idx].placements:
+                            expected_input_specs = output_strategy_list[
+                                idx
+                            ].input_specs or [output_strategy_list[idx].output_spec]
+                            if all(
+                                arg_spec.placements != input_spec.placements
+                                for input_spec in expected_input_specs
+                            ):
+                                if torch.distributed.get_rank() == 0:
+                                    print(
+                                        f"needs_redistribute because arg_spec {arg_spec.placements} != expected_input_specs: {expected_input_specs}"
+                                    )
                                 needs_redistribute = True
-                            expected_input_spec_list.append(out_spec_list[idx])
+                            expected_input_spec_list.append(expected_input_specs[0])
```

Comment:
I think this needs to take an input spec that matched the `arg_spec`, which might not be the 0th one.